src/Conversion/ONNXToKrnl/Tensor/Transpose.cpp

```diff
@@ -46,32 +46,29 @@ struct ONNXTransposeOpLowering : public ConversionPattern {
 
     // Insert an allocation and deallocation for the result of this operation.
     Value alloc = insertAllocAndDeallocSimple(
-        rewriter, op, memRefType, loc, shapeHelper.dimsForOutput(0));
-
-    // Create loop.
-    BuildKrnlLoop inputLoops(rewriter, loc, rank);
-    inputLoops.createDefineAndIterateOp(data);
-    rewriter.setInsertionPointToStart(inputLoops.getIterateBlock());
-    {
-      // Get a child IndexExpr context.
-      IndexExprScope childScope(&rewriter, shapeHelper.scope);
-      KrnlBuilder createKrnl(rewriter, loc);
-
-      // Get read/write indices.
-      SmallVector<IndexExpr, 4> readIndices;
-      SmallVector<IndexExpr, 4> writeIndices;
-      for (decltype(rank) i = 0; i < rank; ++i) {
-        Value readVal = inputLoops.getInductionVar(i);
-        Value writeVal =
-            inputLoops.getInductionVar(ArrayAttrIntVal(permAttr, i));
-        readIndices.emplace_back(DimIndexExpr(readVal));
-        writeIndices.emplace_back(DimIndexExpr(writeVal));
-      }
-
-      // Copy data.
-      Value loadData = createKrnl.loadIE(data, readIndices);
-      createKrnl.storeIE(loadData, alloc, writeIndices);
-    }
+        rewriter, op, memRefType, loc, shapeHelper.dimsForOutput());
+
+    KrnlBuilder createKrnl(rewriter, loc);
+    ValueRange loopDef = createKrnl.defineLoops(rank);
+    SmallVector<IndexExpr, 4> lbs(rank, LiteralIndexExpr(0));
+
+    MemRefBoundsIndexCapture dataBounds(data);
+    SmallVector<IndexExpr, 4> ubs;
+    for (uint64_t i = 0; i < rank; ++i)
+      ubs.emplace_back(dataBounds.getDim(i));
```

Comment:
This loop can be replaced by `dataBounds.getList(ubs)`.

Reply:
Thanks for pointing that out. Will use `dataBounds.getDimList(ubs)` because `getList` is private.